src/java.base/share/classes/sun/launcher/LauncherHelper.java

```diff
@@ -918,59 +918,35 @@ static void validateMainClass(Class<?> mainClass) {
             // invalid main or not FX application, abort with an error
-            abort(null, "java.launcher.cls.error4", mainClass.getName(),
+            abort(null, "java.launcher.cls.error2", mainClass.getName(),
                   JAVAFX_APPLICATION_CLASS_NAME);
         } catch (Throwable e) {
             if (mainClass.getModule().isNamed()) {
-                abort(e, "java.launcher.module.error5",
+                abort(e, "java.launcher.module.error3",
                       mainClass.getName(), mainClass.getModule().getName(),
                       e.getClass().getName(), e.getLocalizedMessage());
             } else {
-                abort(e, "java.launcher.cls.error7", mainClass.getName(),
+                abort(e, "java.launcher.cls.error5", mainClass.getName(),
                       e.getClass().getName(), e.getLocalizedMessage());
             }
         }
 
         setMainType(mainMethod);
 
-        /*
-         * findMainMethod (above) will choose the correct method, based
-         * on its name and parameter type, however, we still have to
-         * ensure that the method is static (non-preview) and returns a void.
-         */
-        int mods = mainMethod.getModifiers();
-        boolean isStatic = Modifier.isStatic(mods);
-        boolean isPublic = Modifier.isPublic(mods);
-        boolean noArgs = mainMethod.getParameterCount() == 0;
-
-        if (!PreviewFeatures.isEnabled()) {
-            if (!isStatic || !isPublic || noArgs) {
-                abort(null, "java.launcher.cls.error2", "static",
-                      mainMethod.getDeclaringClass().getName());
-            }
-        }
-
-        if (!isStatic) {
+        if (!Modifier.isStatic(mainMethod.getModifiers())) {
+            String className = mainMethod.getDeclaringClass().getName();
             if (mainClass.isMemberClass() && !Modifier.isStatic(mainClass.getModifiers())) {
-                abort(null, "java.launcher.cls.error9",
-                        mainMethod.getDeclaringClass().getName());
+                abort(null, "java.launcher.cls.error7", className);
             }
             try {
                 Constructor<?> constructor = mainClass.getDeclaredConstructor();
                 if (Modifier.isPrivate(constructor.getModifiers())) {
-                    abort(null, "java.launcher.cls.error8",
-                          mainMethod.getDeclaringClass().getName());
+                    abort(null, "java.launcher.cls.error6", className);
                 }
             } catch (Throwable ex) {
-                abort(null, "java.launcher.cls.error8",
-                      mainMethod.getDeclaringClass().getName());
+                abort(null, "java.launcher.cls.error6", className);
             }
         }
-
-        if (mainMethod.getReturnType() != java.lang.Void.TYPE) {
-            abort(null, "java.launcher.cls.error3",
-                  mainMethod.getDeclaringClass().getName());
-        }
```

Comment:
The simplification to LauncherHelper.validateMainClass looks okay but one thing is a bit annoying is side effect to set mainType. What would you think about have validateMainClass return the Method for the main class and have checkAndLoadMain set mainType after validateMainClass succeeds, right not it gets set even if the validate fails.

Reply:
Agree. Will fetch flags from second call from JVM.

Reply:
Amazingly enough I seem to have the JNI straight on the first try.

Reply:
> Amazingly enough I seem to have the JNI straight on the first try.

Oh, I wasn't expecting that. My comment was suggesting we change LauncherHelper.checkAndLoadMain to set both appClass and mainType after the validate succeeds.

Reply:
I thought it was a clearer way to get rid of the side effect.

Reply:
> I thought it was a clearer way to get rid of the side effect.

It means LauncherHelper.checkAndLoadMain description will need to be updated as it says that it checks the public static void main method. But okay, I'll go through what you have now.